### DISCUSSION TITLE
How to calculate this sysex string 12 bit word with negative value?

### DISCUSSION THREAD

**Discussion by jojogib** [RomanKubiak/ctrlr#515] [Q&A]
Hello everyone,
**I want to control the filter and it works for cutoff with value from 0-150**

```
filterenvelope = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

  local bi = BigInteger(value)

  local MS = bi:getBitRangeAsInt(6, 12)
  local LS = bi:getBitRangeAsInt(0, 6)

  -- Send the MIDI message
  
	local mStr = string.format("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 %.2X %.2X F7", MS, LS)
  	panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

end
```

**this works, but for for the filter envelope I need values from -99 to +99
I dont understand how to calculate.**

I managed to to get it worked like this but this takes forever to write:

```
filterenvelope = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
  -- Define the header and end bytes of the SysEx message

if value == -99 then
local mStr = string.format("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 02 01 F7")
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

elseif value == -98 then
local mStr = string.format("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 02 02 F7")
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

elseif value == -97 then
local mStr = string.format("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 02 03 F7")
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

elseif value == -96 then
local mStr = string.format("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 02 05 F7")
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

elseif value == -95 then
local mStr = string.format("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 02 06 F7")
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

elseif value == -94 then
local mStr = string.format("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 02 07 F7")
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

elseif value == -93 then
local mStr = string.format("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 02 09 F7")
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

elseif value == -92 then
local mStr = string.format("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 02 0a F7")
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

elseif value == -91 then
local mStr = string.format("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 02 0b F7")
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

```

Im very bad in coding this is copy paste code I manged to rewrite. May some one could help me.

**Comment by damiensellier** [comment]
Hi,
You're dealing with -/+ range value split on 2 bytes it's very common. 
Are you sure the MSB is not 7F (127) instead of 02 if the value is negative? for ex h7F h01 for value -1 or h7F h02 for -2...
Most manufacturer do -/+ value range split on 2 bytes with MSB 64 to 127 for negative values and 0 to 63 included for positive values. Or even signed 2's complement for smaller range.

**Reply by damiensellier** [reply]
If ever the way you said is the right way, 
with valueMSB = 00 if value is >= 0 
and valueMSB = 02 if value is < 0 
then you just need to write this statement : 


```
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
myMethod = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

console(tostring(value))

if value  >=  0 then  -- for 0 and positive values
	valueMSB = 0x00
	console ("value is +")
	valueLSB = math.abs(value)

else  -- for negative values
	valueMSB = 0x02
	console ("value is -")
	valueLSB = 100 - math.abs(value)
end

console("valueMSB : "..valueMSB)
console("valueLSB : "..valueLSB)

mStr = string.format("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 %02x %02x F7", valueMSB, valueLSB)
panel:sendMidiMessageNow(CtrlrMidiMessage(mStr))

end

```
EDIT: lsb value for neg are 100-abs(modvalue) according to OP, which seems strange. please double check both MSB and LSB values...

**Comment by dnaldoog** [comment]
Hi @jojogib 

What is the machine? Some sort of Ensoniq?

**Comment by dnaldoog** [comment]
I remember @DRBAUS was struggling with this on an **Ensoniq** machine.

I think the MSB changes arbitrarily every 50 values, which is strange, unless there's some logic there I can't see.

And also because the LSB values don't increase linearly, it's best to use a lookup table, which should  be easier to manage than 199 if/else statements:

Declare a table like this:

<pre>
ensoniq_neg_lookup={
257	, --	-99
258	, --	-98
259	, --	-97
261	, --	-96
262	, --	-95
263	, --	-94
265	, --	-93
266	, --	-92
267	, --	-91
268	, --	-90
270	, --	-89
271	, --	-88
272	, --	-87
273	, --	-86
275	, --	-85
276	, --	-84
277	, --	-83
279	, --	-82
280	, --	-81
281	, --	-80
282	, --	-79
284	, --	-78
285	, --	-77
286	, --	-76
288	, --	-75
289	, --	-74
290	, --	-73
291	, --	-72
293	, --	-71
294	, --	-70
295	, --	-69
297	, --	-68
298	, --	-67
299	, --	-66
300	, --	-65
302	, --	-64
303	, --	-63
304	, --	-62
306	, --	-61
307	, --	-60
308	, --	-59
309	, --	-58
311	, --	-57
312	, --	-56
313	, --	-55
314	, --	-54
316	, --	-53
317	, --	-52
318	, --	-51
384	, --	-50
385	, --	-49
386	, --	-48
387	, --	-47
389	, --	-46
390	, --	-45
391	, --	-44
393	, --	-43
394	, --	-42
395	, --	-41
396	, --	-40
398	, --	-39
399	, --	-38
400	, --	-37
401	, --	-36
403	, --	-35
404	, --	-34
405	, --	-33
407	, --	-32
408	, --	-31
409	, --	-30
410	, --	-29
412	, --	-28
413	, --	-27
414	, --	-26
416	, --	-25
417	, --	-24
418	, --	-23
419	, --	-22
421	, --	-21
422	, --	-20
423	, --	-19
425	, --	-18
426	, --	-17
427	, --	-16
428	, --	-15
430	, --	-14
431	, --	-13
432	, --	-12
434	, --	-11
435	, --	-10
436	, --	-9
437	, --	-8
439	, --	-7
440	, --	-6
441	, --	-5
442	, --	-4
444	, --	-3
445	, --	-2
446	, --	-1
0	, --	0
1	, --	1
3	, --	2
4	, --	3
5	, --	4
7	, --	5
8	, --	6
9	, --	7
10	, --	8
12	, --	9
13	, --	10
14	, --	11
15	, --	12
17	, --	13
18	, --	14
19	, --	15
21	, --	16
22	, --	17
23	, --	18
24	, --	19
26	, --	20
27	, --	21
28	, --	22
30	, --	23
31	, --	24
32	, --	25
33	, --	26
35	, --	27
36	, --	28
37	, --	29
39	, --	30
40	, --	31
41	, --	32
42	, --	33
44	, --	34
45	, --	35
46	, --	36
48	, --	37
49	, --	38
50	, --	39
51	, --	40
53	, --	41
54	, --	42
55	, --	43
56	, --	44
58	, --	45
59	, --	46
60	, --	47
62	, --	48
63	, --	49
128	, --	50
129	, --	51
131	, --	52
132	, --	53
133	, --	54
135	, --	55
136	, --	56
137	, --	57
138	, --	58
140	, --	59
141	, --	60
142	, --	61
143	, --	62
145	, --	63
146	, --	64
147	, --	65
149	, --	66
150	, --	67
151	, --	68
152	, --	69
154	, --	70
155	, --	71
156	, --	72
158	, --	73
159	, --	74
160	, --	75
161	, --	76
163	, --	77
164	, --	78
165	, --	79
167	, --	80
168	, --	81
169	, --	82
170	, --	83
172	, --	84
173	, --	85
174	, --	86
176	, --	87
177	, --	88
178	, --	89
179	, --	90
181	, --	91
182	, --	92
183	, --	93
184	, --	94
186	, --	95
187	, --	96
188	, --	97
190	, --	98
191	, --	99
}
</pre>

and parse the MSB,LSB table value using _BigInteger_ and send as MIDI:

<pre>
calc_filter = function(mod, value, source)
local index=value+100 -- a value of -99 will access table index [1]
local m=MemoryBlock("f0 0f 03 00 11 00 00 00 00 00 01 14 02 00 00 00 00 F7")
  local bi = BigInteger(ensoniq_neg_lookup[index])
  local MS = bi:getBitRangeAsInt(7, 7)
  local LS = bi:getBitRangeAsInt(0, 6)
	m:setByte(15,MS)
	m:setByte(16,LS)
  	panel:sendMidiMessageNow(CtrlrMidiMessage(m:toHexString(1)))
end
</pre>

Of course this is all a wild guess, so please check the values carefully! Panel attached.

![ensoniq](https://user-images.githubusercontent.com/76831683/208910149-499fa2ba-f73e-4df3-b6a0-dde70b0c49d8.png)

[Discussion 515 ensoniq_2_0.zip](https://github.com/RomanKubiak/ctrlr/files/10277803/Discussion.515.ensoniq_2_0.zip)

**Reply by damiensellier** [reply]
Yes it's for an Ensoniq, it has 0F for manufacturer code in the header.
61,62,63,65??? Wow those value increments don't make any sense :)
The OP will be happy with that, you're right, it's the perfect solution John.

**Reply by dnaldoog** [reply]
It's strange, isn't it?

I just looked up the code in a panel @DRBAUS had sent me when he was trying to work out the code for that. I then changed the code to get rid of all the if /else statements.

I questioned those numbers, but he assured me they were correct. 😄

**Reply by damiensellier** [reply]
With Ensoniq everything is possible :)

**Reply by jojogib** [reply]
just tested it. works great. Thank you :)

**Comment by DRBAUS** [comment]
Seeing the sysex I can tell it's for the EPS-16 Plus or ASR-10.

I like your way of thinking. I am not good at coding either and actually started my panel with a line for every value for every parameter....and it works....but not for every parameter (sample start/end/length etc). 

I have come a long way already in with this panel but got frustrated and have not been working on it for about a year now....I think. Time is needed for me to work it out....and I don't have much time to work on this because of other things that demand my time and attention. 

Ensoniq is weird when it comes to sysex but we (the people in this discussion) have come a long way in streamlining the code.

Hopefully in the near future I will be able to get back to programming.

**Reply by jojogib** [reply]
I feel your pain. Some parameter were easy like cutoff.. But loke you said sample start and end is a different beast

**Comment by jojogib** [comment]
Thank you for all your help.
Yeah its for an asr-10.
I managed to get sample start and end to work almost. But now I´m stucked.

```
change = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
  local myValue  = nil
  local myHexVal = nil

  -- Get the hexadecimal string from the "uiLabelText" property of the "hans3" component
  myValue = panel:getComponent("hans3"):getProperty("uiLabelText")

  -- Create a MIDI message object from the hexadecimal string
  myHexVal = CtrlrMidiMessage(myValue)

  if modulator == 1 then

    -- Set the values of bytes 363 through 369 based on the value of the modulator
    myHexVal:setByte(363, 0x00)
    myHexVal:setByte(364, 0x00)
    myHexVal:setByte(365, 0x00)
    myHexVal:setByte(366, 0x1c)
    myHexVal:setByte(367, 0x00)
    myHexVal:setByte(368, 0x09)
    myHexVal:setByte(369, 0x38)
  
  elseif modulator == 0 then

    -- Set the values of bytes 363 through 369 based on the value of the modulator
    myHexVal:setByte(363, 0x00)
    myHexVal:setByte(364, 0x00)
    myHexVal:setByte(365, 0x00)
    myHexVal:setByte(366, 0x00)
    myHexVal:setByte(367, 0x00)
    myHexVal:setByte(368, 0x00)
    myHexVal:setByte(369, 0x08)
  end

  
  -- Send the "myHexVal" MIDI message
  panel:sendMidiMessageNow(myHexVal)
end

```

I wanna take the Hex-string from the UI-Lable and change diffrent bytes based on the modulator value. The method I wrote dont work. Maybe someone have an idea.

**Reply by dnaldoog** [reply]
Hi @jojogib 

Can you post the panel here, so we can see what you are trying to achieve?

**Comment by damiensellier** [comment]
` if modulator == 1 then`

_modulator_ is not previously defined locally, is it it stored from somwhere else? Or did you mean _value_ or _myValue_ instead, for a modulator whose _value_ switches between 0 or 1?

**Comment by jojogib** [comment]
[uitext2string.zip](https://github.com/RomanKubiak/ctrlr/files/10308852/uitext2string.zip)

I dont know if its more clear now what i want to archive

**Comment by dnaldoog** [comment]
This is the correct way to do it, but you __cannot__ send a value of `0xff` in MIDI in one byte. The limit is `0x7f`!

<pre>
changeSysexByte = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    if panel:getBootstrapState() then
        return
    end -- stop this function from firing when panel loads for the first time
    -- Get the hexadecimal string from the "uiLabelText" property of the "sysex" component
    local myValue = L(panel:getLabelComponent("sysex"):getText())
    -- Create a MIDI message object from the hexadecimal string
    local m = MemoryBlock()
    m:loadFromHexString(myValue)
    -- Check the value of the "changeSysexByte" modulator
    if value == 1 then
        -- Set the value of byte 3 to 0xff
        m:setByte(3, 0xff) -- needs to be split into 2 bytes!!
    elseif value == 2 then
        -- Set the value of byte 2 to 0xff
        m:setByte(2, 0xff) -- needs to be split into 2 bytes!!
    end
    panel:sendMidiMessageNow(CtrlrMidiMessage(m:toHexString(1)))
end
</pre>

Your code:

<pre>
changeSysexByte = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
  
  local myValue  = nil -- unnecessary initialization
  local myHexVal = nil-- unnecessary initialization
  local myModulator = panel:getModulatorByName("changeSysexByte") -- unnecessary THIS == value (function argument passed in above)

  -- Get the hexadecimal string from the "uiLabelText" property of the "sysex" component
  myValue = panel:getComponent("sysex"):getProperty("uiLabelText") -- needs L() ??

  -- Create a MIDI message object from the hexadecimal string
  myHexVal = CtrlrMidiMessage(myValue) 
  
  -- Check the value of the "changeSysexByte" modulator
  if myModulator:getValue() == 1 then

    -- Set the value of byte 3 to 0xff
    myHexVal:setByte(3, 0xff) -- will crash panel
-- MIDI cannot be 0xff  &#x2021;
--myHex variable reference to CtrlrMidiMessage() cannot take non existent setByte() method which only belongs to MemoryBlock()!!!

  elseif myModulator:getValue() == 2 then

    -- Set the value of byte 2 to 0xff
    myHexVal:setByte(2, 0xff) -- will crash panel
  end

  -- Send the modified "myHexVal" MIDI message
  panel:sendMidiMessageNow(myHexVal)
end
</pre>

> &#x2021; Yeah exactly, if it's like other Ensoniq products, you have to "nibblize" the bytes by pair where 0xff is 0x0f 0x0f. This is the way Ensoniq sysex messages are usually sent for higher values than 0x7f.  ✔

**Reply by damiensellier** [reply]
Yeah exactly, if it's like other Ensoniq products, you have to "nibblize" the bytes by pair where **0xff** is **0x0f 0x0f**. This is the way Ensoniq sysex messages are usually sent for higher values than 0x7f. 
On the DP4 the  parameter values where split on a total of 4 bytes (max 65536 values), where **VV vv** is **0V 0V 0v 0v** to comply to the midi sysex format. I don't know if it's working this way for the ASR10 though.

Can you please copy the link to your formula in the sysex manual for this particular command?
